File: 서지수/Week_9/P92341_주차요금계산.py
def solution(fees, records):
    cars={}
    # 나간 시간이 없는 경우 탐색 후 OUT을 추가
    for i in range(len(records)):
        if records[i].split(" ")[1] in cars:
            cars[records[i].split(" ")[1]]=int(cars[records[i].split(" ")[1]])+1
        else:
            cars[records[i].split(" ")[1]]=1
    for i in cars.keys():
        for j in  range(len(records)):
            if records[j].split(" ")[0]=="23:59":
                records.remove(records[j])
            elif int(i)!=2:
                records.append("23:59 "+i+" OUT")
    
    cars_times={}
    # 차 시간 구하기
    for i in cars.keys():
        cars_times[i]=[]
    
    for i in range(len(records)):
        if records[i].split(" ")[1] in cars.keys():
            cars_times[records[i].split(" ")[1]].append(records[i].split(" ")[0])
    # 차 번호 순으로 정렬
    cars_time=dict(sorted(cars_times.items()))

    # 요금 매기기
    cars_money={}
    # 차 시간 구하기
    for i in cars.keys():
        cars_money[i]=[]

    for i in cars_time.keys():
        time=int(cars_time[i][1].split(":")[0]+cars_time[i][1].split(":")[1])-int(cars_time[i][0].split(":")[0]+cars_time[i][0].split(":")[1])
        hour=time//100
        minute=time%100
        if hour*60+minute<=fees[0]:
            cars_money[i]=fees[1]
        else:
            cars_money[i]=fees[1]+(((hour-fees[0//60])*60+minute))//fees[2]*fees[3]

    return list(cars_money.values())

File: 윤상우/Week_9/P92341_주차요금계산.java
package temp;

import java.util.*;

public class P92341_주차요금계산 {
    public int[] solution(int[] fees, String[] records) {
        HashMap<Integer, Integer> map = new HashMap<>();
        HashMap<Integer, Integer> map_cal = new HashMap<>();
        List<Integer> list = new ArrayList<>();

        for(String r : records){
            int num = Integer.parseInt(r.substring(6,10));
            if(!map.containsKey(num)){
                map.put(num,0);
                list.add(num);
            }
        }

        for(String r : records){
            int hour = Integer.parseInt(r.substring(0,2));
            int minute = Integer.parseInt(r.substring(3,5));
            int num = Integer.parseInt(r.substring(6,10));
            String record = r.substring(11);

            int time = hour * 60 + minute;

            if(record.equals("IN")){
                map_cal.put(num, time);
            }else{
                int temp = map_cal.get(num);
                map_cal.remove(num);
                map.put(num, map.get(num) + time - temp);
            }
        }

        list.sort((o1,o2) -> o1 - o2);

        int[] answer = new int [list.size()];

        for(int i : list){
            if(map_cal.containsKey(i)){
                map.put(i, map.get(i) + 1439 - map_cal.get(i));
            }
        }

        for(int i=0; i<list.size(); i++){
            int money = 0;
            int time = map.get(list.get(i));
            if(time > fees[0]){
                if((time - fees[0]) % fees[2] > 0){
                    money += fees[1] + ((time - fees[0]) / fees[2] + 1)  * fees[3];
                }else{
                    money += fees[1] + ((time - fees[0]) / fees[2]) * fees[3];
                }
            }else{
                money += fees[1];
            }

            answer[i] = money;

        }

        System.out.println(map);

        return answer;
    }
}